src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/service/external/VerifyEmailServiceTest.kt

```diff
@@ -0,0 +1,240 @@
+    }
+
+    @Test
+    fun shouldFormatEmailInput() {
+      val user = createSampleUserWith(username = "someuser@user.com", email = "testy@testing.com")
+      whenever(authApiService.createTokenByEmailType(TokenByEmailTypeRequest(user.username, EmailType.PRIMARY.name))).thenReturn(token)
+      whenever(verifyEmailDomainService.isValidEmailDomain("somewhere.com")).thenReturn(true)
+
+      val linkEmailAndUsername = verifyEmailService.requestVerification(user, "    some.u’ser@SOMEwhere.COM", url)
+
+      assertThat(linkEmailAndUsername.email).isEqualTo("some.u'ser@somewhere.com")
+    }
+  }
+
+  @Nested
+  inner class ValidateEmailAddress {
+
+    @Test
+    fun gsiEmail() {
+      whenever(verifyEmailDomainService.isValidEmailDomain(anyString())).thenReturn(true)
+      verifyPrimaryEmailFailure("some.u'ser@SOMEwhe.gsi.gov.uk", "gsi")
+    }
+
+    @Test
+    fun noAtSign() {
+      verifyPrimaryEmailFailure("a", "format")
+    }
+
+    @Test
+    fun multipleAtSigns() {
+      verifyPrimaryEmailFailure("a@b.fred@joe.com", "at")
+    }
+
+    @Test
+    fun noExtension() {
+      verifyPrimaryEmailFailure("a@bee", "format")
+    }
+
+    @Test
+    fun firstLastStopFirst() {
+      verifyPrimaryEmailFailure(".a@bee.com", "firstlast")
+    }
+
+    @Test
+    fun firstLastStopLast() {
+      verifyPrimaryEmailFailure("a@bee.com.", "firstlast")
+    }
+
+    @Test
+    fun firstLastAtFirst() {
+      verifyPrimaryEmailFailure("@a@bee.com", "firstlast")
+    }
+
+    @Test
+    fun firstLastAtLast() {
+      verifyPrimaryEmailFailure("a@bee.com@", "firstlast")
+    }
+
+    @Test
+    fun togetherAtBefore() {
+      verifyPrimaryEmailFailure("a@.com", "together")
+    }
+
+    @Test
+    fun togetherAtAfter() {
+      verifyPrimaryEmailFailure("a.@joe.com", "together")
+    }
+
+    @Test
+    fun white() {
+      verifyPrimaryEmailFailure("a@be\te.com", "white")
+    }
+
+    @Test
+    fun invalidCharacters() {
+      verifyPrimaryEmailFailure("a@b.&com", "characters")
+    }
+
+    @Test
+    fun invalidDomain() {
+      verifyPrimaryEmailFailure("a@b.com", "domain")
+      verify(verifyEmailDomainService).isValidEmailDomain("b.com")
+    }
+
+    @Test
+    fun `validate email exceeds email max length`() {
+      val email: String = "A".repeat(241)
+      assertThatThrownBy {
+        verifyEmailService.validateEmailAddress(
+          email,
+          EmailType.PRIMARY
+        )
+      }.isInstanceOf(VerifyEmailService.ValidEmailException::class.java)
+        .hasMessage("Validate email failed with reason: maxlength")
+    }
+
+    private fun verifyPrimaryEmailFailure(email: String, reason: String) {
+      assertThatThrownBy { verifyEmailService.validateEmailAddress(email, EmailType.PRIMARY) }.isInstanceOf(
+        VerifyEmailService.ValidEmailException::class.java
+      ).extracting("reason").isEqualTo(reason)
+    }
+  }
+
+  private fun createSampleUserWith(username: String, email: String): ExternalUserDetailsDto {
```

Comment:
Should prob pull this out into a Helper class for everybody to use for testing

Reply:
Done